pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py

```diff
@@ -0,0 +1,546 @@
+
+            PREDS.extend(pred)
+            TARGETS.extend(target)
+
+            if args.rank == 0:
+                print(
+                    "Val epoch {}/{} {}/{}".format(epoch, args.epochs, idx, len(loader)),
+                    "loss: {:.4f}".format(loss),
+                    "acc: {:.4f}".format(acc),
+                    "time {:.2f}s".format(time.time() - start_time),
+                )
+            start_time = time.time()
+
+        # Calculate QWK metric (Quadratic Weigted Kappa) https://en.wikipedia.org/wiki/Cohen%27s_kappa
+        PREDS = PREDS.get_buffer().cpu().numpy()
+        TARGETS = TARGETS.get_buffer().cpu().numpy()
+        qwk = cohen_kappa_score(PREDS.astype(np.float64), TARGETS.astype(np.float64), weights="quadratic")
+
+    return loss, acc, qwk
+
+
+def save_checkpoint(model, epoch, args, filename="model.pt", best_acc=0):
+    """Save checkpoint"""
+
+    state_dict = model.state_dict() if not args.distributed else model.module.state_dict()
+
+    save_dict = {"epoch": epoch, "best_acc": best_acc, "state_dict": state_dict}
+
+    filename = os.path.join(args.logdir, filename)
+    torch.save(save_dict, filename)
+    print("Saving checkpoint", filename)
+
+
+class LabelEncodeIntegerGraded(Transform):
+    """
+    Convert an integer label to encoded array representation of length num_classes,
+    with 1 filled in up to label index, and 0 otherwise. For example for num_classes=5,
+    embedding of 2 -> (1,1,0,0,0)
+
+    Args:
+        num_classes: the number of classes to convert to encoded format.
+        keys: keys of the corresponding items to be transformed
+            Defaults to ``['label']``.
+
+    """
+
+    def __init__(self, num_classes, keys=["label"]):
+        super().__init__()
+        self.keys = keys
+        self.num_classes = num_classes
+
+    def __call__(self, data):
+
+        d = dict(data)
+        for key in self.keys:
+            label = int(d[key])
+
+            lz = np.zeros(self.num_classes, dtype=np.float32)
+            lz[:label] = 1.0
+            # alternative oneliner lz=(np.arange(self.num_classes)<int(label)).astype(np.float32) #same oneliner
+            d[key] = lz
+
+        return d
+
+
+def main():
+
+    args = parse_args()
+
+    if args.dataset_json is None:
+        # download default json datalist
+        resource = "https://drive.google.com/uc?id=1L6PtKBlHHyUgTE4rVhRuOLTQKgD4tBRK"
+        dst = "./datalist_panda_0.json"
+        if not os.path.exists(dst):
+            gdown.download(resource, dst, quiet=False)
+        args.dataset_json = dst
+
+    if args.distributed:
+        ngpus_per_node = torch.cuda.device_count()
+        args.optim_lr = ngpus_per_node * args.optim_lr / 2  # heuristic to scale up learning rate in multigpu setup
+        args.world_size = ngpus_per_node * args.world_size
+
+        print("Multigpu", ngpus_per_node, "rescaled lr", args.optim_lr)
+        mp.spawn(main_worker, nprocs=ngpus_per_node, args=(args,))
```

Comment:
I remember `torch.distributed.launch` can provide better performance for multi-GPU training, should we change it?
CC @vfdev-5 

Thanks.

Reply:
I do not have measurements right now, but I remember checking that some time ago and in recent pytorch versions the difference is almost invisible. I think you can keep it.

Reply:
Hi @vfdev-5 , thanks for your confirmation.

Reply:
sounds like we can keep it